src/security/builtin_plugins/cryptographic/src/crypto_transform.c

```diff
@@ -745,38 +748,45 @@ add_specific_mac(
 {
-  uint32_t index;
-  const uint8_t prefix_kind = is_rtps ? SMID_SRTPS_PREFIX : SMID_SEC_PREFIX;
-  const uint8_t postfix_kind = is_rtps ? SMID_SRTPS_POSTFIX : SMID_SEC_POSTFIX;
-  trusted_crypto_data_t data;
-  struct trusted_crypto_header *header;
-  struct trusted_crypto_footer *footer;
-  crypto_session_key_t key;
-  crypto_hmac_t hmac;
-
-  trusted_crypto_buffer_append(buffer, sizeof(struct receiver_specific_mac));
-
-  if ((header = trusted_crypto_buffer_find_header(buffer, prefix_kind)) == NULL)
+  size_t header_offset, footer_offset;
+  if (!add_specific_mac_find_offsets (buffer, is_rtps, &header_offset, &footer_offset))
     return false;
-  else if ((footer = trusted_crypto_buffer_find_footer(buffer, postfix_kind)) == NULL)
-    return false;
-
-  footer->header.octetsToNextHeader = (uint16_t)(footer->header.octetsToNextHeader + sizeof(struct receiver_specific_mac));
-  index = ddsrt_fromBE4u(footer->postfix.receiver_specific_macs._length);
-  data.x.base = footer->postfix.common_mac.data;
-  data.x.length = CRYPTO_HMAC_SIZE;
 
-  if (!crypto_calculate_receiver_specific_key(&key, session->id, keymat->master_salt, keymat->master_receiver_specific_key, keymat->transformation_kind, ex) ||
-      !crypto_cipher_encrypt_data(&key, session->key_size, &header->prefix.iv, 1, &data, NULL, &hmac, ex))
-    return false;
+  crypto_hmac_t hmac;
+  {
+    struct trusted_crypto_header const * const h = (struct trusted_crypto_header const *) (buffer->contents + header_offset);
+    struct trusted_crypto_footer const * const f = (struct trusted_crypto_footer const *) (buffer->contents + footer_offset);
+    crypto_session_key_t key;
+    const trusted_crypto_data_t data = { {
+      .base = (unsigned char *) f->postfix.common_mac.data,
+      .length = CRYPTO_HMAC_SIZE
+    } };
+    if (!crypto_calculate_receiver_specific_key (&key, session->id, keymat->master_salt, keymat->master_receiver_specific_key, keymat->transformation_kind, ex) ||
+        !crypto_cipher_encrypt_data (&key, session->key_size, &h->prefix.iv, 1, &data, NULL, &hmac, ex))
+      return false;
+  }
 
-  uint32_t key_id = ddsrt_toBE4u(keymat->receiver_specific_key_id);
-  struct trusted_crypto_postfix *postfix = &footer->postfix;
-  struct receiver_specific_mac *rcvmac = &postfix->receiver_specific_macs._buffer[index];
+  // appending may force reallocation
+  trusted_crypto_buffer_append (buffer, sizeof (struct receiver_specific_mac));
+  struct trusted_crypto_footer * const footer = (struct trusted_crypto_footer *) (buffer->contents + footer_offset);
+  const uint32_t index = ddsrt_fromBE4u (footer->postfix.receiver_specific_macs._length);
+
+  // Coverity gets upset by using a byteswapped value without checking it on the assumption that
+  // byteswapping implies external input ...
+  const size_t receiver_specific_macs_offset = CRYPTO_HMAC_SIZE + 4;
+  if (index > (footer->header.octetsToNextHeader - receiver_specific_macs_offset) / sizeof (struct receiver_specific_mac))
```

Comment:
Maybe use `length` instead of `index` here, because this is validating the current list of macs, not if there is room for one more